integration/integration_test_base.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import os
+import socket
+import subprocess
+import sys
+import threading
+import time
+import unittest
+
+from common import IpVersion
+from nighthawk_test_server import NighthawkTestServer
+
+
+class IntegrationTestBase(unittest.TestCase):
+  """
+    IntegrationTestBase facilitates testing against the Nighthawk test server, by running
+    determining a free port, and starting it up in a separate process in setUp().
+    """
+
+  def __init__(self, *args, **kwargs):
+    super(IntegrationTestBase, self).__init__(*args, **kwargs)
+    self.test_rundir = os.path.join(os.environ["TEST_SRCDIR"], os.environ["TEST_WORKSPACE"])
+
+    self.nighthawk_test_server_path = os.path.join(self.test_rundir, "nighthawk_test_server")
+    self.nighthawk_test_config_path = None
+    self.nighthawk_client_path = os.path.join(self.test_rundir, "nighthawk_client")
+    self.server_ip = "::1" if IntegrationTestBase.ip_version == IpVersion.ipv6 else "127.0.0.1"
+    self.socket_type = socket.AF_INET6 if IntegrationTestBase.ip_version == IpVersion.ipv6 else socket.AF_INET
+    self.test_server = None
+    self.server_port = -1
+    self.parameters = dict()
+
+  def getFreeListenerPortForAddress(self, address):
+    """
+        Determines a free port and returns that. Theoretically it is possible that another process
+        will steal the port before our caller is able to leverage it, but we take that chance.
+        The upside is that we can push the port upon the server we are about to start through configuration
+        which is compatible accross servers.
+        """
+    sock = socket.socket(self.socket_type, socket.SOCK_STREAM)
+    sock.bind((address, 0))
+    port = sock.getsockname()[1]
+    sock.close()
+    return port
+
+  def setUp(self):
+    """
+        Performs sanity checks and starts up the server. Upon exit the server is ready to accept connections.
+        """
+    self.assertTrue(os.path.exists(self.nighthawk_test_server_path))
+    self.assertTrue(os.path.exists(self.nighthawk_client_path))
+    self.server_port = self.getFreeListenerPortForAddress(self.server_ip)
+    self.test_server = NighthawkTestServer(
+        self.nighthawk_test_server_path, self.nighthawk_test_config_path, self.server_ip,
+        self.server_port, IntegrationTestBase.ip_version, self.parameters)
+    self.assertTrue(self.test_server.start())
+
+  def tearDown(self):
+    """
+        Stops the server.
+        """
+    self.assertEqual(0, self.test_server.stop())
+
+  def getNighthawkCounterMapFromJson(self, parsed_json):
+    """
+        Utility method to get the counters from the json indexed by name.
+        Note:the `client.` prefix will be stripped from the index.
+        """
+    counter_map = {}
+    counters = parsed_json["results"][0]["counters"]
+    for counter in counters:
+      counter_qualified_name = counter["name"]
+      counter_map[counter_qualified_name[len("client."):]] = int(counter["value"])
+    return counter_map
```

Comment:
This could even be more concise as a Python dictionary comprehension.

Reply:
done